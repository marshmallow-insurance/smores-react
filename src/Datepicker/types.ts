export interface Day {
  label: string
  disabled: boolean
  active: boolean
  date: Date
}
