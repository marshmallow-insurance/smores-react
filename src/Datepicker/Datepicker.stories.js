import React from 'react'
import { Datepicker } from './Datepicker'

export default {
  title: 'Datepicker',
  component: Datepicker,
  argTypes: { onDateSelect: { action: 'date selected' } },
}

const Template = (args) => <Datepicker {...args} />

export const Default = Template.bind({})

Default.args = {
  disableWeekend: false,
  firstDayShift: true,
  range: 31,
  fromDate: new Date('2021/07/14'),
}
