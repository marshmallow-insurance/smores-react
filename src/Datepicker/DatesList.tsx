import React, { FC } from 'react'
import styled from 'styled-components'
import { getISODay } from 'date-fns'

import { theme } from '../theme'

import { Day } from './types'

type Props = {
  items: Day[]
  handleDateSelect: (date: Date) => void
}

const getBlankDaysCount = (firstDayOfTheMonth: Date) => {
  const dayOfTheWeek = getISODay(firstDayOfTheMonth)
  return dayOfTheWeek - 1
}

export const DatesList: FC<Props> = ({ items, handleDateSelect }) => {
  return (
    <Container>
      {Array(getBlankDaysCount(items[0].date))
        .fill(null)
        .map((_, index) => (
          <ListItem key={`blankItem-${index}`} disabled />
        ))}
      {items.map((item: Day, i) => (
        <ListItem
          key={i}
          disabled={item.disabled}
          className={`ListItem ${item.active ? 'active' : ''}`}
          onClick={() => handleDateSelect(item.date)}
          data-testid={item.label}
        >
          {item.label}
        </ListItem>
      ))}
    </Container>
  )
}

const Container = styled.ul`
  display: grid;
  grid-column-gap: 4px;
  grid-row-gap: 8px;
  grid-template-columns: repeat(7, 1fr);
  padding: 8px;

  @media (min-width: 360px) {
    grid-gap: 8px;
    padding: 16px;
  }

  @media (min-width: 768px) {
    grid-column-gap: 16px;
    padding: 32px;
  }
`

interface IListItem {
  disabled?: boolean
}

const ListItem = styled.li<IListItem>`
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 4px;
  width: 32px;
  height: 32px;
  border: 1px solid transparent;
  font-weight: ${theme.font.weight.medium};
  font-size: 14px;
  line-height: 15px;
  color: ${theme.colors.secondary};
  background-color: transparent;
  pointer-events: ${({ disabled }) => (disabled ? 'none' : '')};
  opacity: ${({ disabled }) => (disabled ? '0.2' : '1')};
  cursor: pointer;

  &:hover:not(.active) {
    background-color: ${theme.colors.background};
  }

  &.active {
    color: ${theme.colors.white};
    background-color: ${theme.colors.secondary};
    cursor: default;
  }
`
