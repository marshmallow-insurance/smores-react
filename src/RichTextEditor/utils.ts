import { $createParagraphNode, LexicalNode, RootNode } from 'lexical'

export const appendNodes = (root: RootNode, nodes: LexicalNode[]) => {
  nodes
    .filter((node) => node.__type !== 'linebreak')
    .map((node) => {
      if (node.__type === 'text') {
        const paragraphNode = $createParagraphNode()
        paragraphNode.append(node)
        return paragraphNode
      }
      return node
    })
    .forEach((node) => root.append(node))
}
