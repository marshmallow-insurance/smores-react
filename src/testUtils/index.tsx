import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
