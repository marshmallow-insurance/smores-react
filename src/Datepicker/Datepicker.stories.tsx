import React from 'react'
import { Datepicker, DatepickerProps } from './Datepicker'

export default {
  title: 'Datepicker',
  component: Datepicker,
  argTypes: { onDateSelect: { action: 'date selected' } },
}

const Template = (props: DatepickerProps) => <Datepicker {...props} />

export const Default = Template.bind({})

Default.args = {
  showDayLabels: false,
  disableWeekend: false,
  firstDayShift: true,
  range: 31,
  fromDate: new Date('2021/07/14'),
}
