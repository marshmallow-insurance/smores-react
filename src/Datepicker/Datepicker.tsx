import {
  addDays,
  eachMonthOfInterval,
  format,
  getDaysInMonth,
  getMonth,
  getYear,
  isSameDay,
  isSameMonth,
  isWeekend,
  isWithinInterval,
} from 'date-fns'
import React, { FC, useState } from 'react'
import styled, { css } from 'styled-components'

import { Box } from '../Box'
import { Icon } from '../Icon'
import { Text } from '../Text'
import { theme } from '../theme'

import { convertToUkDate } from '../utils/date'
import { focusOutlineStyle } from '../utils/focusOutline'
import { MarginProps } from '../utils/space'
import { useControllableState } from '../utils/useControlledState'
import { DatesList } from './DatesList'

const getAvailableMonths = (startDate: Date, endDate: Date) => {
  const monthList = eachMonthOfInterval({
    start: startDate,
    end: endDate,
  })

  return monthList.map((monthDate) => ({
    date: monthDate,
    label: format(monthDate, 'MMMM'),
  }))
}

export type DatepickerProps = {
  showDayLabels?: boolean
  disableWeekend?: boolean
  fromDate?: Date
  endingDate?: Date
  showYear?: boolean
  showSelectedDate?: boolean
  range?: number
  onDateSelect: (date: string) => void
  onChange?: (value: Date) => void
  value?: Date
  /** Disables all dates, defaults to false */
  disabled?: boolean
  fallbackStyle?: boolean
} & MarginProps

export const Datepicker: FC<DatepickerProps> = ({
  showDayLabels = false,
  disableWeekend = true,
  range = 14,
  fromDate = new Date(),
  endingDate,
  showYear = false,
  showSelectedDate = false,
  onDateSelect,
  onChange,
  value,
  disabled,
  fallbackStyle = false,
  ...marginProps
}) => {
  // We want to make sure that the date is in the UK timezone,
  // this might need to be revisit when opening up to new countries
  const startDate = convertToUkDate(fromDate)
  const endDate = endingDate ? endingDate : addDays(startDate, range)
  const availableMonths = getAvailableMonths(startDate, endDate)

  const selectedDate = value ?? new Date()

  const [activeDay, setActiveDay] = useControllableState({
    initialState: undefined,
    stateProp: value,
  })
  const [activeMonthIndex, setActiveMonth] = useState(
    showSelectedDate
      ? availableMonths.findIndex((month) =>
          isSameMonth(month.date, selectedDate),
        )
      : 0,
  )

  const handleSelectEvent = (date: Date) => {
    setActiveDay(date)
    onChange?.(date)
    onDateSelect(format(date, 'yyyy-MM-dd'))
  }

  const generateDaysForMonth = (monthDate: Date) => {
    const daysInMonth = getDaysInMonth(monthDate)
    const month = getMonth(monthDate)
    const year = getYear(monthDate)
    const filteredDays = []

    for (let i = 1; i <= daysInMonth; i += 1) {
      const date = new Date(year, month, i)

      filteredDays.push({
        date,
        label: format(date, 'dd'),
        active: activeDay ? isSameDay(date, activeDay) : false,
        disabled: disabled
          ? true
          : !isWithinInterval(date, { start: startDate, end: endDate }) ||
            (disableWeekend && isWeekend(date)),
      })
    }

    return filteredDays
  }

  return (
    <Container id="datepicker" {...marginProps} $fallbackStyle={fallbackStyle}>
      <Header
        flex
        alignItems="center"
        justifyContent="space-between"
        direction="row"
      >
        <Circle
          aria-label="previous-month"
          type="button"
          data-testid="last-month"
          disabled={activeMonthIndex === 0}
          onClick={() => setActiveMonth(activeMonthIndex - 1)}
          onKeyDown={(e) => {
            if (e.key === 'Enter') {
              setActiveMonth(activeMonthIndex - 1)
            }
          }}
        >
          <Icon render="caret" color="cream" size={18} rotate={90} />
        </Circle>

        <Heading tag="h4" typo="body-regular">
          {availableMonths[activeMonthIndex]?.label}{' '}
          {showYear && `- ${getYear(availableMonths[activeMonthIndex]?.date)}`}
        </Heading>

        <Circle
          aria-label="next-month"
          type="button"
          data-testid="next-month"
          disabled={activeMonthIndex === availableMonths.length - 1}
          onClick={() => setActiveMonth(activeMonthIndex + 1)}
          onKeyDown={(e) => {
            if (e.key === 'Enter') {
              setActiveMonth(activeMonthIndex + 1)
            }
          }}
        >
          <Icon render="caret" color="cream" size={18} rotate={-90} />
        </Circle>
      </Header>

      <Box flex alignItems="center" justifyContent="center">
        <DatesList
          items={generateDaysForMonth(availableMonths[activeMonthIndex]?.date)}
          showDayLabels={showDayLabels}
          handleDateSelect={handleSelectEvent}
        />
      </Box>
    </Container>
  )
}

const Container = styled(Box)<{ $fallbackStyle: boolean }>`
  font-family: ${theme.font.system};
  display: inline-block;
  box-sizing: border-box;
  outline: none;
  border-radius: 16px;

  ${({ $fallbackStyle, theme }) => {
    return $fallbackStyle
      ? css`
          background-color: ${theme.color.surface.base['000']};
        `
      : css`
          background-color: ${theme.color.surface.base[300]};
        `
  }}
`

const Header = styled(Box)`
  color: ${({ theme }) => theme.color.text.base};
  user-select: none;
  padding: 16px;
  border-bottom: 1px solid ${({ theme }) => theme.color.border.subtle};

  span {
    font-size: 16px;
    line-height: 1;
  }
`

const Heading = styled(Text)`
  position: relative;
  top: 2px;
  font-weight: ${theme.font.weight.medium};
`

const Circle = styled.button`
  background-color: ${({ theme }) => theme.color.surface.base[900]};
  height: 32px;
  width: 32px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  border: none;
  padding: 0 !important;
  cursor: pointer;

  :disabled {
    opacity: 0.5;
    cursor: default;
  }

  ${focusOutlineStyle}
`
