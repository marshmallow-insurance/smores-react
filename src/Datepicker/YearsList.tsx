import React, { FC } from 'react'
import styled from 'styled-components'

import { theme } from '../theme'

import { focusOutlineStyle } from '../utils/focusOutline'

type Props = {
  handleYearSelect: (year: number) => void
  selectedYear?: number
}

export const YearsList: FC<Props> = ({ handleYearSelect, selectedYear }) => {
  const years = Array.from(Array(200).keys()).map((year) => year + 1900)

  return (
    <Container>
      {years.map((year, i) => (
        <ListButton
          key={i}
          type="button"
          $active={selectedYear === year}
          onClick={() => handleYearSelect(year)}
        >
          {year}
        </ListButton>
      ))}
    </Container>
  )
}

const Container = styled.div`
  justify-items: center;
  display: grid;
  grid-column-gap: 4px;
  grid-row-gap: 8px;
  grid-template-columns: repeat(4, 1fr);
  padding: 8px;
  height: 208px;
  margin-bottom: 16px;
  overflow: scroll;
  width: 304px;

  @media (min-width: 360px) {
    grid-gap: 8px;
    padding: 16px;
  }

  @media (min-width: 768px) {
    grid-column-gap: 16px;
    padding: 32px;
  }
`

interface IListButton {
  disabled?: boolean
  $active?: boolean
}

const ListButton = styled.button<IListButton>`
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 8px;
  width: 50px;
  height: 32px;
  border: 1px solid transparent;
  font-weight: ${theme.font.weight.medium};
  font-size: 14px;
  line-height: 15px;
  color: ${theme.colors.liquorice};
  background-color: transparent;
  pointer-events: ${({ disabled }) => (disabled ? 'none' : '')};
  opacity: ${({ disabled }) => (disabled ? '0.2' : '1')};
  cursor: pointer;

  &:hover:not(.active) {
    color: ${theme.colors.custard};
    background-color: ${theme.colors.liquorice};
  }

  ${({ $active }) =>
    $active &&
    `color: ${theme.colors.custard};
    background-color: ${theme.colors.liquorice};
    cursor: default;`}

  &.active {
    color: ${theme.colors.custard};
    background-color: ${theme.colors.liquorice};
    cursor: default;
  }

  ${focusOutlineStyle}
`
