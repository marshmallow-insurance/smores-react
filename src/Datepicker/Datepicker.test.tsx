import React from 'react'
import { expect, it } from 'vitest'
import { render } from '../testUtils'
import { Datepicker } from './Datepicker'
import { noop } from '../utils/noop'

describe('Datepicker', () => {
  it('renders correctly with default props', () => {
    const { container } = render(
      <Datepicker
        onDateSelect={noop}
        fromDate={new Date('2020-07-01')}
        range={14}
      />,
    )
    expect(container).toMatchSnapshot()
  })
})
