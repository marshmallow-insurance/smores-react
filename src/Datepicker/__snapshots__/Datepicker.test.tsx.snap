// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Datepicker > renders correctly with default props 1`] = `
<div>
  <div
    class="sc-bRKDuR bHjpxP sc-jJLAfE bTWFLo"
    id="datepicker"
  >
    <div
      class="sc-bRKDuR hsryNT sc-hwkwBN kshGKm"
    >
      <button
        aria-label="previous-month"
        class="sc-dYwGCk eoezIS"
        data-testid="last-month"
        disabled=""
        type="button"
      >
        <span
          class="sc-bRKDuR bHjpxP sc-fhHczv bvDwHH"
          color="cream"
          data-testid="caret-container"
          rotate="90"
          size="18"
        >
          <svg
            fill="none"
            height="100%"
            viewBox="0 0 24 24"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.66 8.248a1 1 0 00-1.412.095L12 15.482l-6.248-7.14a1 1 0 10-1.504 1.317l7 8a.995.995 0 001.504 0l7-8a1 1 0 00-.093-1.411z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
      <h4
        class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT sc-kNOymR kGWfzP"
        cursor="inherit"
        title=""
      >
        July
         
      </h4>
      <button
        aria-label="next-month"
        class="sc-dYwGCk eoezIS"
        data-testid="next-month"
        disabled=""
        type="button"
      >
        <span
          class="sc-bRKDuR bHjpxP sc-fhHczv uyYpS"
          color="cream"
          data-testid="caret-container"
          rotate="-90"
          size="18"
        >
          <svg
            fill="none"
            height="100%"
            viewBox="0 0 24 24"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.66 8.248a1 1 0 00-1.412.095L12 15.482l-6.248-7.14a1 1 0 10-1.504 1.317l7 8a.995.995 0 001.504 0l7-8a1 1 0 00-.093-1.411z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="sc-bRKDuR fwCHSf"
    >
      <div
        class="sc-jwTyAe lnXCrR"
      >
        <button
          aria-label="0-unselectable-date"
          class="sc-hjsuWn dXBqJk"
          disabled=""
        />
        <button
          aria-label="1-unselectable-date"
          class="sc-hjsuWn dXBqJk"
          disabled=""
        />
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="01"
          type="button"
        >
          01
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="02"
          type="button"
        >
          02
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="03"
          type="button"
        >
          03
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="04"
          disabled=""
          type="button"
        >
          04
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="05"
          disabled=""
          type="button"
        >
          05
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="06"
          type="button"
        >
          06
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="07"
          type="button"
        >
          07
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="08"
          type="button"
        >
          08
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="09"
          type="button"
        >
          09
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="10"
          type="button"
        >
          10
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="11"
          disabled=""
          type="button"
        >
          11
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="12"
          disabled=""
          type="button"
        >
          12
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="13"
          type="button"
        >
          13
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="14"
          type="button"
        >
          14
        </button>
        <button
          class="sc-hjsuWn fsGFMP ListButton "
          data-testid="15"
          type="button"
        >
          15
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="16"
          disabled=""
          type="button"
        >
          16
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="17"
          disabled=""
          type="button"
        >
          17
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="18"
          disabled=""
          type="button"
        >
          18
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="19"
          disabled=""
          type="button"
        >
          19
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="20"
          disabled=""
          type="button"
        >
          20
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="21"
          disabled=""
          type="button"
        >
          21
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="22"
          disabled=""
          type="button"
        >
          22
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="23"
          disabled=""
          type="button"
        >
          23
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="24"
          disabled=""
          type="button"
        >
          24
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="25"
          disabled=""
          type="button"
        >
          25
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="26"
          disabled=""
          type="button"
        >
          26
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="27"
          disabled=""
          type="button"
        >
          27
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="28"
          disabled=""
          type="button"
        >
          28
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="29"
          disabled=""
          type="button"
        >
          29
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="30"
          disabled=""
          type="button"
        >
          30
        </button>
        <button
          class="sc-hjsuWn dXBqJk ListButton "
          data-testid="31"
          disabled=""
          type="button"
        >
          31
        </button>
      </div>
    </div>
  </div>
</div>
`;
